File: codes/cpp/chap08/answer_A57.cpp
#include <iostream>
using namespace std;

int N, Q, A[100009], X[100009], Y[100009];
int dp[32][100009];

int main() {
	// 入力
	cin >> N >> Q;
	for (int i = 1; i <= N; i++) cin >> A[i];
	for (int j = 1; j <= Q; j++) cin >> X[j] >> Y[j];

	// 前計算
	for (int i = 1; i <= N; i++) dp[0][i] = A[i];
	for (int d = 1; d <= 29; d++) {
		for (int i = 1; i <= N; i++) dp[d][i] = dp[d - 1][dp[d - 1][i]];
	}

	// クエリの処理
	for (int j = 1; j <= Q; j++) {
		int CurrentPlace = X[j];
		for (int d = 29; d >= 0; d--) {
			if ((Y[j] / (1 << d)) % 2 != 0) CurrentPlace = dp[d][CurrentPlace];
		}
		cout << CurrentPlace << endl;
	}
	return 0;
}

File: codes/python/chap08/answer_A57.py
# 入力
N, Q = map(int, input().split())
A = list(map(int, input().split()))
queries = [ list(map(int, input().split())) for i in range(Q) ]

# 前計算（ここでは書籍とは異なり、0-indexed で実装しています）
LEVELS = 30
dp = [ [ None ] * N for i in range(LEVELS) ]
for i in range(0, N):
	dp[0][i] = A[i] - 1
for d in range(1, LEVELS):
	for i in range(0, N):
		dp[d][i] = dp[d - 1][dp[d - 1][i]]

# クエリの処理（ここでは書籍とは異なり、0-indexed で実装しています）
for X, Y in queries:
	current_place = X - 1
	for d in range(29, -1, -1):
		if ((Y >> d) & 1) == 1:
			current_place = dp[d][current_place]
	print(current_place + 1) # current_place は 0-indexed で計算したので、1 を足して出力する

# 注意 1：書籍の C++ プログラムにおいて「Y の 2^d の位を取り出す」は C++ では (Y / (1 << d)) % 2 と実装されていました。
#         Python でも 19 行目は (Y // (2 ** d)) % 2 と実装できますが、((Y >> d) & 1) と計算すると、定数倍の面でより高速です。
# 注意 2：このプログラムの平均的な実行時間は 2 秒にギリギリ入るくらいであり、提出によっては TLE となる場合があります。
#         同じプログラムを PyPy3 で提出すると、0.5 秒程度の実行時間で AC することができます。